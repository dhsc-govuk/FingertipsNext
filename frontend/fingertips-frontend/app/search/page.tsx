import { SearchForm } from '@/components/forms/SearchForm';

export default async function Page(
  props: Readonly<{
    searchParams?: Promise<{
      indicator?: string;
    }>;
  }>
) {
  const searchParams = await props.searchParams;
  const indicator = searchParams?.indicator ?? '';

  return <SearchForm indicator={indicator} />;
}
