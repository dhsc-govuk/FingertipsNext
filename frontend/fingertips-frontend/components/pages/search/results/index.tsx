'use client';

import { BackLink, H1, Paragraph } from 'govuk-react';

export function SearchResults({ indicator }: Readonly<{ indicator: string }>) {
  return (
    <>
      <BackLink href={`/search?indicator=${indicator}`} />
      <H1>Search results</H1>
      <Paragraph>{`You searched for indicator "**${indicator}**"`}</Paragraph>
    </>
  );
}
