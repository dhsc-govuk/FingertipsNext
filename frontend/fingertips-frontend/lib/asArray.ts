export const asArray = (value?: string | string[]): string[] => {
  if (value) {
    return Array.isArray(value) ? value : [value];
  }
  return [];
};
