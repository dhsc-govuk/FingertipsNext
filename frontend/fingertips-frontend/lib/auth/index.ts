import { AuthConfigFactory } from '@/lib/auth/config';
import NextAuth from 'next-auth';

export const { handlers, signIn, signOut, auth } = NextAuth(
  AuthConfigFactory.getConfig()
);
