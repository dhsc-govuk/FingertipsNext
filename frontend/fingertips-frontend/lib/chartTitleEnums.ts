export enum ChartTitlesEnum {
  LineChart = 'Indicator trends over time',
  BarChartEmbeddedTable = 'Compare areas for one time period',
  Heatmap = 'Overview of indicators and areas',
  SpineChart = 'Area profile by indicators',
  InequalitiesBarChart = 'Inequalities comparison for one time period',
  InequalitiesLineChart = 'Inequalities trends over time',
  ThematicMap = 'Compare an indicator by areas',
  PopulationPyramid = 'Population pyramid',
}
