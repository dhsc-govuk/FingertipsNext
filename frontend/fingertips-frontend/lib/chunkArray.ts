export const maxNumAreasThatCanBeRequestedAPI = 300;

export function chunkArray(
  arrayToChunk: string[],
  chunkSize: number = maxNumAreasThatCanBeRequestedAPI
) {
  const chunkedArray = [];
  for (let i = 0; i < arrayToChunk.length; i += chunkSize) {
    chunkedArray.push(arrayToChunk.slice(i, i + chunkSize));
  }
  return chunkedArray;
}
