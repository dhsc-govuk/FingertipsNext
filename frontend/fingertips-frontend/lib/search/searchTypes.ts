/*********************************************************************************
 * This file includes duplicate definitions from the search-setup/src/constants.ts
 * These should be updated to take definitions from there rather than duplicating them
 */

export const INDICATOR_SEARCH_SCORING_PROFILE = 'basicScoringProfile';
export const INDICATOR_SEARCH_INDEX_NAME = 'indicator-search-index';
export const AREA_SEARCH_INDEX_NAME = 'area-search-index';
export const AREA_SEARCH_SUGGESTER_NAME = 'areaSuggester';

export type IndicatorDocument = {
  indicatorId: string;
  name: string;
  definition: string;
  dataSource: string;
  latestDataPeriod: string;
  lastUpdated: Date;
};

export type AreaDocument = {
  areaCode: string;
  areaType: string;
  areaName: string;
};

export interface IIndicatorSearchService {
  searchWith(searchTerm: string): Promise<IndicatorDocument[]>;
}

export interface IAreaSearchService {
  getAreaSuggestions(partialAreaName: string): Promise<AreaDocument[]>;
}
