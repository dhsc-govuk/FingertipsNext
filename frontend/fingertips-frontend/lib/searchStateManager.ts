import { asArray } from './pageHelpers';

export enum SearchParams {
  SearchedIndicator = 'si',
  IndicatorsSelected = 'is',
  AreasSelected = 'as',
  AreaTypeSelected = 'ats',
  GroupTypeSelected = 'gts',
  ConfidenceIntervalSelected = 'cis',
}

export type SearchParamKeys = `${SearchParams}`;

const multiValueParams = [
  SearchParams.IndicatorsSelected as string,
  SearchParams.AreasSelected as string,
  SearchParams.ConfidenceIntervalSelected as string,
];

export type SearchStateParams = {
  [SearchParams.SearchedIndicator]?: string;
  [SearchParams.IndicatorsSelected]?: string[];
  [SearchParams.AreasSelected]?: string[];
  [SearchParams.AreaTypeSelected]?: string;
  [SearchParams.GroupTypeSelected]?: string;
  [SearchParams.ConfidenceIntervalSelected]?: string[];
};

const isMultiValueTypeParam = (searchParamKey: SearchParamKeys) =>
  multiValueParams.includes(searchParamKey);

export class SearchStateManager {
  private searchState: SearchStateParams;
  private searchStateParams: URLSearchParams;

  constructor(searchState: SearchStateParams = {}) {
    this.searchState = {};
    this.searchStateParams = new URLSearchParams();

    this.searchState = searchState;
  }

  private constructPath(path: string) {
    if (this.searchStateParams.size === 0) {
      return path;
    }
    return `${path}?${this.searchStateParams.toString()}`;
  }

  private addParamToPath(searchParamKey: SearchParamKeys) {
    const paramsToAdd = asArray(this.searchState[searchParamKey]);

    if (paramsToAdd.length > 0) {
      paramsToAdd.forEach((paramToAdd) => {
        this.searchStateParams.append(searchParamKey, paramToAdd);
      });
    }
  }

  public addParamValueToState(
    searchParamKey: SearchParamKeys,
    paramToAdd: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = this.searchState[searchParamKey];

      const currentParamStateAsArray = asArray(currentParamState);
      const doesParamExist = currentParamStateAsArray.find(
        (currentParam) => currentParam === paramToAdd
      );

      if (!doesParamExist) {
        const newState: SearchStateParams = {
          ...this.searchState,
          [searchParamKey]: [...currentParamStateAsArray, paramToAdd],
        };

        this.searchState = newState;
      }
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: paramToAdd,
      };

      this.searchState = newState;
    }
  }

  public removeParamValueFromState(
    searchParamKey: SearchParamKeys,
    paramToRemove?: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = asArray(this.searchState[searchParamKey]);

      const newParamState = currentParamState?.filter((paramValue) => {
        return paramValue !== paramToRemove;
      });

      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: newParamState,
      };

      this.searchState = newState;
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: undefined,
      };

      this.searchState = newState;
    }
  }

  public removeAllParamFromState(searchParamKey: SearchParamKeys) {
    this.searchState[searchParamKey] = undefined;
  }

  public static setStateFromParams(params: URLSearchParams) {
    const newState = Object.values(SearchParams).reduce<SearchStateParams>(
      (state, searchParamKey) => {
        if (isMultiValueTypeParam(searchParamKey)) {
          const paramValues = params.getAll(searchParamKey);

          if (paramValues.length > 0) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValues,
            };

            return newState;
          }
        } else {
          const paramValue = params.get(searchParamKey);

          if (paramValue) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValue,
            };

            return newState;
          }
        }
        return state;
      },
      {}
    );

    const searchStateManager = new SearchStateManager(newState);
    return searchStateManager;
  }

  public static setStateFromSearchStateParams(params?: SearchStateParams) {
    if (params) {
      const newState = Object.values(SearchParams).reduce<SearchStateParams>(
        (state, searchParamKey) => {
          if (isMultiValueTypeParam(searchParamKey)) {
            const paramValues = asArray(params[searchParamKey]);

            if (paramValues.length > 0) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValues,
              };

              return newState;
            }
          } else {
            const paramValue = params[searchParamKey];

            if (paramValue) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValue,
              };

              return newState;
            }
          }
          return state;
        },
        {}
      );

      const searchStateManager = new SearchStateManager(newState);
      return searchStateManager;
    }
    return new SearchStateManager();
  }

  public getSearchState() {
    return this.searchState;
  }

  public generatePath(path: string) {
    this.searchStateParams = new URLSearchParams();

    Object.values(SearchParams).forEach((searchParamKey) => {
      this.addParamToPath(searchParamKey);
    });

    return this.constructPath(path);
  }
}
