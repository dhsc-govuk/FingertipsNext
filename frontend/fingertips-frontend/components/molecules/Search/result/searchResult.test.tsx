import { render, screen } from '@testing-library/react';
import { expect } from '@jest/globals';
import { SearchResult } from '.';
import { userEvent } from '@testing-library/user-event';
import { SearchParams } from '@/lib/searchStateManager';
import { IndicatorSearchResult } from '@/lib/search/searchTypes';

const mockPath = 'some-mock-path';
const mockReplace = jest.fn();

jest.mock('next/navigation', () => {
  const originalModule = jest.requireActual('next/navigation');

  return {
    ...originalModule,
    usePathname: () => mockPath,
    useSearchParams: () => ({ [SearchParams.SearchedIndicator]: 'test' }),
    useRouter: jest.fn().mockImplementation(() => ({
      replace: mockReplace,
    })),
  };
});

const dummyIndicatorResult: IndicatorSearchResult = {
  indicatorId: '1',
  name: 'NHS',
  definition: 'Total number of patients registered with the practice',
  latestDataPeriod: '2023',
  dataSource: 'NHS website',
  lastUpdated: new Date('December 6, 2024'),
};

afterEach(() => {
  mockReplace.mockClear();
});

it('should have search result list item', () => {
  render(<SearchResult result={dummyIndicatorResult} />);

  expect(screen.getByRole('listitem')).toBeInTheDocument();
});

it('should contain 3 paragraphs and a heading', () => {
  render(<SearchResult result={dummyIndicatorResult} />);

  expect(screen.getAllByRole('paragraph')).toHaveLength(3);
  expect(screen.getByRole('heading')).toBeInTheDocument();
});

it('should contain expected text', () => {
  render(<SearchResult result={dummyIndicatorResult} />);

  expect(screen.getByRole('heading').textContent).toContain('NHS');
  expect(screen.getAllByRole('paragraph').at(0)?.textContent).toContain('2023');
  expect(screen.getAllByRole('paragraph').at(1)?.textContent).toContain(
    'NHS website'
  );
  expect(screen.getAllByRole('paragraph').at(2)?.textContent).toContain(
    '06 December 2024'
  );
});

describe('Indicator Checkbox', () => {
  it('should mark the checkbox as checked if indicatorSelected is true', () => {
    render(
      <SearchResult result={dummyIndicatorResult} indicatorSelected={true} />
    );

    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('should mark the checkbox as not checked if indicatorSelected is false', () => {
    render(
      <SearchResult result={dummyIndicatorResult} indicatorSelected={false} />
    );

    expect(screen.getByRole('checkbox')).not.toBeChecked();
  });

  it('should update the path when an indicator is checked', async () => {
    const user = userEvent.setup();
    mockReplace.mockClear();

    render(
      <SearchResult result={dummyIndicatorResult} indicatorSelected={false} />
    );

    await user.click(screen.getByRole('checkbox'));

    expect(mockReplace).toBeCalledWith(
      `${mockPath}?${SearchParams.SearchedIndicator}=test&${SearchParams.IndicatorsSelected}=1`,
      {
        scroll: false,
      }
    );
  });

  it('should update the path when an indicator is unchecked', async () => {
    const user = userEvent.setup();

    render(
      <SearchResult result={dummyIndicatorResult} indicatorSelected={true} />
    );

    await user.click(screen.getByRole('checkbox'));

    expect(mockReplace).toBeCalledWith(
      `${mockPath}?${SearchParams.SearchedIndicator}=test`,
      {
        scroll: false,
      }
    );
  });

  it('should have a direct link to the indicator chart', () => {
    const expectedPath = `/chart?${SearchParams.SearchedIndicator}=test&${SearchParams.IndicatorsSelected}=${dummyIndicatorResult.indicatorId.toString()}`;
    render(<SearchResult result={dummyIndicatorResult} />);

    expect(screen.getByRole('link')).toHaveAttribute('href', expectedPath);
  });

  it('snapshot test', () => {
    const container = render(<SearchResult result={dummyIndicatorResult} />);

    expect(container.asFragment()).toMatchSnapshot();
  });
});
