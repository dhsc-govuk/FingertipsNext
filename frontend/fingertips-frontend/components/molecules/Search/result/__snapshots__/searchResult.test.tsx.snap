// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Result Suite snapshot test 1`] = `
<DocumentFragment>
  .c0 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 5px;
}

.c4 {
  color: #0b0c0c;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 700;
  font-size: 16px;
  line-height: 1.25;
  display: block;
  margin-top: 0;
  margin-bottom: 15px;
}

.c5 {
  margin: 0;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 15px;
}

.c5 >p {
  margin: 0;
}

.c5 >p>code {
  padding: 0.2em 0.4em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(27,31,35,0.05);
  border-radius: 3px;
}

.c5 >pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f6f8fa;
  border-radius: 3px;
}

.c5 >pre>code {
  display: inline;
  padding: 0;
  margin: 0;
  border: 0;
  overflow: visible;
  line-height: inherit;
  word-wrap: normal;
}

.c3 {
  box-sizing: border-box;
  padding-right: 15px;
  padding-left: 15px;
}

.c1 {
  display: flex;
  flex-direction: column;
  margin: 0 -15px;
}

.c7 {
  margin: 0;
  border: 0;
  border-bottom: 1px solid #bfc1c3;
}

.c6 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 0.5;
}

.c2 {
  padding-top: 15px;
  padding-bottom: 0;
  padding-left: 25px;
  padding-right: 25px;
}

@media print {
  .c0 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width: 641px) {
  .c0 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c4 {
    color: #000;
  }
}

@media print {
  .c4 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width: 641px) {
  .c4 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width: 641px) {
  .c4 {
    margin-bottom: 20px;
  }
}

@media print {
  .c5 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width: 641px) {
  .c5 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width: 641px) {
  .c5 {
    margin-bottom: 20px;
  }
}

@media only screen and (min-width: 641px) {
  .c3 {
    flex-grow: 1;
    flex-shrink: 1;
  }
}

@media only screen and (min-width: 641px) {
  .c1 {
    flex-direction: row;
  }
}

@media print {
  .c6 {
    font-size: 14px;
    line-height: 0.5;
  }
}

@media only screen and (min-width: 641px) {
  .c6 {
    font-size: 19px;
    line-height: 0.5;
  }
}

@media only screen and (min-width: 641px) {
  .c2 {
    padding-top: 15px;
  }
}

@media only screen and (min-width: 641px) {
  .c2 {
    padding-bottom: 0;
  }
}

@media only screen and (min-width: 641px) {
  .c2 {
    padding-left: 40px;
  }
}

@media only screen and (min-width: 641px) {
  .c2 {
    padding-right: 40px;
  }
}

<li
    class="c0"
    data-testid="search-result"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <h5
          class="c4"
        >
          NHS
        </h5>
        <div
          class="c5 c6"
        >
          <p>
            Latest data period: 2023
          </p>
        </div>
        <div
          class="c5 c6"
        >
          <p>
            Data source: NHS website
          </p>
        </div>
        <div
          class="c5 c6"
        >
          <p>
            Last updated: 06 December 2024
          </p>
        </div>
      </div>
    </div>
    <hr
      class="c7"
    />
  </li>
</DocumentFragment>
`;
